From ae05f495f5d41dff09c7e4cfed059bf9ecf43fc2 Mon Sep 17 00:00:00 2001
From: Rex Lu <rex.lu@mediatek.com>
Date: Tue, 7 Oct 2025 17:19:21 +0800
Subject: [PATCH 013/105] mtk: mt76: mt7996: fix extender mode call trace
 during L1 SER in hw scan

2. add sanity check on mt76_worker_disable to avoid set tx_worker to park state twice
and happend kernel waring "pc : kthread_park+0x9c/0xb0".

Refactor sanity check on mt76_worker_disable part.
previous way still have chance hit kernel warning

Signed-off-by: Rex Lu <rex.lu@mediatek.com>
Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 util.h | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/util.h b/util.h
index b3b9393cc..5e995ff20 100644
--- a/util.h
+++ b/util.h
@@ -21,6 +21,7 @@ struct mt76_worker
 enum {
 	MT76_WORKER_SCHEDULED,
 	MT76_WORKER_RUNNING,
+	MT76_WORKER_DISABLED,
 };
 
 #define MT76_INCR(_var, _size) \
@@ -95,8 +96,11 @@ static inline void mt76_worker_disable(struct mt76_worker *w)
 	if (!w->task)
 		return;
 
+	if (test_and_set_bit(MT76_WORKER_DISABLED, &w->state))
+		return;
+
 	kthread_park(w->task);
-	WRITE_ONCE(w->state, 0);
+	WRITE_ONCE(w->state, BIT(MT76_WORKER_DISABLED));
 }
 
 static inline void mt76_worker_enable(struct mt76_worker *w)
@@ -105,6 +109,7 @@ static inline void mt76_worker_enable(struct mt76_worker *w)
 		return;
 
 	kthread_unpark(w->task);
+	clear_bit(MT76_WORKER_DISABLED, &w->state);
 	mt76_worker_schedule(w);
 }
 
-- 
2.45.2

